FSE: Integrate wp-env environment

noahtallen opened the pull request:
Originally, I covered several more things in this PR which I have split out into the following:
- [Fix JS unit tests (ready to go)
](https://github.com/Automattic/wp-calypso/pull/39326)
- [Add phpunit tests (WIP)
](https://github.com/Automattic/wp-calypso/pull/39328)
- [Add e2e tests (WIP)
](https://github.com/Automattic/wp-calypso/pull/39327)

This PR sets up [@wordpress/env](https://github.com/WordPress/gutenberg/tree/master/packages/env) as a Docker environment for running the plugin locally (and hopefully, in the future, in CI).

As part of this effort, I have created an opinionated "setup-env" script which sets up the repositories a developer needs for working with the FSE plugin locally.
- It clones gutenberg and themes as siblings of wp-calypso in your filesystem if they are not there yet. (So if you have gutenberg, themes, and wp-calypso cloned in the same directory, nothing will change.)
- It builds FSE plugin
- It starts wp-env on `localhost:4013`

### Changes proposed in this Pull Request
- This PR integrates @wordress/env directly into the FSE plugin so that we have a WordPress environment in which to run our tests.
- Adds an environment init script to help set up the local env (includes cloning Gutenberg and themes.)
- Adds a `.wp-env.json` file to specify dependencies of the project.

### Test Install Step
1. In your gutenberg checkout (which should be a sibling of wp-calypso), make sure you have the latest changes on the `master` branch, and build everything: `nvm use && npm ci && npm run build`.
2. From wp-calypso root, run `./apps/full-site-editing/bin/setup-env.sh`. Once it does everything, it will ask you if you want to run the development version of wp-env. **Choose yes since the changes we require have not been published to npm yet.**
3. If all goes well, you should be able to access a Docker instance at `localhost:4013`. (Note: I have experienced issues with my browser caching a redirect between different ports, so you may need to try different browsers and/or clear the cache for `localhost`).

### To do:
- [ ] Check what would be needed to make the setup-env script work in CircleCI
- [x] Ability to run env commands with the correct port numbers ([see my Gutenberg PR](https://github.com/WordPress/gutenberg/pull/20158))

noahtallen commented:
apparently it is impossible to entirely override the ports: https://github.com/docker/compose/issues/2260

noahtallen commented:
Solved by adding the port as an environment variable in the run script :)

jeraldjuice commented on apps/full-site-editing/README.md:
````diff
@@ -9,6 +9,23 @@ This app contains:
+
+```sh
+# run from wp-calypso root. lerna version of `npm run env connect`
+npx lerna run env --scope='@automattic/full-site-editing' --stream -- install
````
❤️

jeraldjuice commented on apps/full-site-editing/package.json:
```diff
@@ -46,5 +47,18 @@
+		"plugin-dir": "full-site-editing-plugin",
+		"plugin-name": "Full Site Editing Plugin",
+		"docker-template": "./bin/docker-compose.override.yml.template",
+		"welcome-logo": [
```
The important part! :)

noahtallen replied:
also, if anyone has ideas on a better ASCII logo, feel free to add it. I just pulled it off an ascii text generator website

jeraldjuice commented on apps/full-site-editing/bin/docker-compose.override.yml.template:
```diff
@@ -0,0 +1,13 @@
+services:
```
nit - can dry this out a little. Depending on your taste for abstracting, with or without a base service:

```
services:
    wordpress-develop:
        &fse-base-service
        volumes:
           - ../full-site-editing-plugin:/var/www/${LOCAL_DIR-src}/wp-content/plugins/%PLUGIN_INSTALL_DIR%

    php:
        <<: *fse-base-service

    cli:
        <<: *fse-base-service

    # etc...
```

noahtallen replied:
Woah!!! That's pretty cool. Will add that :)

noahtallen replied:
This is super helpful too, because we might need to add more volumes for the unit testing stuff

noahtallen commented:
The failing e2e tests don't appear to be related to our additions here. I do see this error:

```
Found console error: "https://wordpress.com/ - A cookie associated with a cross-site resource at http://wordpress.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032." on url 'https://hash-cd9beeca81bec5df19f0cc05faaf2a738983ff08.calypso.live/read'
0
```

noahtallen commented:
Added @Automattic/luna in case folks have opinions on some of this stuff. After all, we'll all be working in this environment :)

chriskmnds commented on apps/full-site-editing/bin/js-unit-config.js:
```diff
@@ -0,0 +1,35 @@
+ *   2. Must have .test.EXT at the end of the filename
+ *   3. EXT (above) must be one of js, ts, jsx, or tsx.
+ * 
+ * Note: In order to use a different jest conig for e2e tests, this config file
```
```suggestion
 * Note: In order to use a different jest config for e2e tests, this config file
```

mattwiebe commented:
After running the `npx lerna run env --scope='@automattic/full-site-editing' --stream -- install` command, I get errors activating the plugin:

```
> env LOCAL_PORT=9999 wp-scripts env "cli" "plugin" "activate" "full-site-editing-plugin"
> WordPress@5.4.0 env:cli /Users/matt/www/calypso/apps/full-site-editing/wordpress
> node ./tools/local-env/scripts/docker.js run cli "--path=/var/www/src" "plugin" "activate" "full-site-editing-plugin"
Plugin 'full-site-editing-plugin' activated.
Success: Activated 1 of 1 plugins.
Warning: filemtime(): stat failed for /var/www/src/wp-content/plugins/full-site-editing-plugin/starter-page-templates/dist/starter-page-templates.js in /var/www/src/wp-content/plugins/full-site-editing-plugin/starter-page-templates/class-starter-page-templates.php on line 72
Warning: require(dist/editor.asset.php): failed to open stream: No such file or directory in /var/www/src/wp-content/plugins/full-site-editing-plugin/blog-posts-block/index.php on line 41
Fatal error: require(): Failed opening required 'dist/editor.asset.php' (include_path='.:/usr/local/lib/php') in /var/www/src/wp-content/plugins/full-site-editing-plugin/blog-posts-block/index.php on line 41
```

Visiting `localhost:9999` results in this:

```
Warning: filemtime(): stat failed for /var/www/src/wp-content/plugins/full-site-editing-plugin/starter-page-templates/dist/starter-page-templates.js in /var/www/src/wp-content/plugins/full-site-editing-plugin/starter-page-templates/class-starter-page-templates.php on line 72

Warning: require(dist/editor.asset.php): failed to open stream: No such file or directory in /var/www/src/wp-content/plugins/full-site-editing-plugin/blog-posts-block/index.php on line 41

Fatal error: require(): Failed opening required 'dist/editor.asset.php' (include_path='.:/usr/local/lib/php') in /var/www/src/wp-content/plugins/full-site-editing-plugin/blog-posts-block/index.php on line 41
```

I'll leave it there for now as I go to debug what went wrong and see if I can help with a fix.

noahtallen commented:
> Warning: filemtime(): stat failed

@mattwiebe I've seen that error before while running the plugin on wpcom, and usually it is fixed by re-building the plugin. It sounds like we may need to run the fse build step _before_ trying to load it in the environment.

noahtallen commented:
@mattwiebe I fixed it by making the script also run the build first.

Try: `npm run distclean` from the root. Then, run `docker system prune -a` to delete any docker images you have. Then try again with this command (slightly different name now that I moved it to a script.)

`npx lerna run init --scope='@automattic/full-site-editing' --stream`

noahtallen commented on .eslintrc.js:
```diff
@@ -33,6 +33,14 @@ module.exports = {
 			},
 		},
+		{
+			files: [ 'apps/full-site-editing/**/tests/**' ],
```
TODO: improve this so that it matches .test.js and .spec.js files

mattwiebe commented:
Okay, passed the first step. Had to run `npm ci` after 'npm run distclean` and before `npx lerna run init --scope='@automattic/full-site-editing' --stream` - we could probably add `npm run install-if-deps-outdated` to the our `init` command to remove the need to remember the extra step.

Also, the first time I ran it and then logged into `wp-admin`, the FSE plugin showed up as deactivated due to not existing, but then the second time I ran it, the plugin is still there and active.

Then I started to try to run the tests.

✅ `test:php`
❌ `test:e2e`
✅  `test:js`

Here's the `e2e` failure output:

```
lerna notice cli v3.20.2
lerna info versioning independent
lerna notice filter including "@automattic/full-site-editing"
lerna info filter [ '@automattic/full-site-editing' ]
lerna info Executing command in 1 package: "npm run test:e2e"
@automattic/full-site-editing: > @automattic/full-site-editing@0.18.2 test:e2e /Users/matt/www/calypso/apps/full-site-editing
@automattic/full-site-editing: > npx wp-scripts test-e2e --wordpress-base-url='http://localhost:9999'
@automattic/full-site-editing: (node:98699) UnhandledPromiseRejectionWarning: Error: Navigation failed because browser has disconnected!
@automattic/full-site-editing:     at CDPSession.<anonymous> (/Users/matt/www/calypso/node_modules/puppeteer/lib/LifecycleWatcher.js:46:107)
@automattic/full-site-editing:     at CDPSession.emit (events.js:210:5)
@automattic/full-site-editing:     at CDPSession._onClosed (/Users/matt/www/calypso/node_modules/puppeteer/lib/Connection.js:215:10)
@automattic/full-site-editing:     at Connection._onMessage (/Users/matt/www/calypso/node_modules/puppeteer/lib/Connection.js:105:17)
@automattic/full-site-editing:     at WebSocket.<anonymous> (/Users/matt/www/calypso/node_modules/puppeteer/lib/WebSocketTransport.js:44:24)
@automattic/full-site-editing:     at WebSocket.onMessage (/Users/matt/www/calypso/node_modules/ws/lib/event-target.js:120:16)
@automattic/full-site-editing:     at WebSocket.emit (events.js:210:5)
@automattic/full-site-editing:     at Receiver.receiverOnMessage (/Users/matt/www/calypso/node_modules/ws/lib/websocket.js:744:20)
@automattic/full-site-editing:     at Receiver.emit (events.js:210:5)
@automattic/full-site-editing:     at Receiver.dataMessage (/Users/matt/www/calypso/node_modules/ws/lib/receiver.js:417:14)
@automattic/full-site-editing:   -- ASYNC --
@automattic/full-site-editing:     at Frame.<anonymous> (/Users/matt/www/calypso/node_modules/puppeteer/lib/helper.js:111:15)
@automattic/full-site-editing:     at Page.waitForNavigation (/Users/matt/www/calypso/node_modules/puppeteer/lib/Page.js:695:49)
@automattic/full-site-editing:     at Page.waitForNavigation (/Users/matt/www/calypso/node_modules/puppeteer/lib/helper.js:112:23)
@automattic/full-site-editing:     at _callee$ (/Users/matt/www/calypso/node_modules/@wordpress/e2e-test-utils/build/@wordpress/e2e-test-utils/src/login-user.js:29:28)
@automattic/full-site-editing:     at tryCatch (/Users/matt/www/calypso/node_modules/regenerator-runtime/runtime.js:45:40)
@automattic/full-site-editing:     at Generator.invoke [as _invoke] (/Users/matt/www/calypso/node_modules/regenerator-runtime/runtime.js:271:22)
@automattic/full-site-editing:     at Generator.prototype.<computed> [as next] (/Users/matt/www/calypso/node_modules/regenerator-runtime/runtime.js:97:21)
@automattic/full-site-editing:     at asyncGeneratorStep (/Users/matt/www/calypso/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
@automattic/full-site-editing:     at _next (/Users/matt/www/calypso/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
@automattic/full-site-editing:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
@automattic/full-site-editing: (node:98699) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
@automattic/full-site-editing: (node:98699) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
@automattic/full-site-editing: FAIL full-site-editing-plugin/full-site-editing/tests/e2e/fse-back-button.spec.js (7.549s)
@automattic/full-site-editing:   Full Site Editing Back Button
@automattic/full-site-editing:     ✕ Should navigate to a new post. (5005ms)
@automattic/full-site-editing:   ● Full Site Editing Back Button › Should navigate to a new post.
@automattic/full-site-editing:     : Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Error:
@automattic/full-site-editing:        6 |
@automattic/full-site-editing:        7 | describe( 'Full Site Editing Back Button', () => {
@automattic/full-site-editing:     >  8 | 	it( 'Should navigate to a new post.', async () => {
@automattic/full-site-editing:          | 	^
@automattic/full-site-editing:        9 | 		await createNewPost();
@automattic/full-site-editing:       10 | 		const title = await page.title();
@automattic/full-site-editing:       11 | 		expect( title ).toBe( 'Add New Post ‹ WordPress Develop — WordPress' );
@automattic/full-site-editing:       at new Spec (../../node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
@automattic/full-site-editing:       at Suite.<anonymous> (full-site-editing-plugin/full-site-editing/tests/e2e/fse-back-button.spec.js:8:2)
@automattic/full-site-editing: Test Suites: 1 failed, 1 total
@automattic/full-site-editing: Tests:       1 failed, 1 total
@automattic/full-site-editing: Snapshots:   0 total
@automattic/full-site-editing: Time:        7.961s, estimated 13s
@automattic/full-site-editing: Ran all test suites.
@automattic/full-site-editing: npm ERR! code ELIFECYCLE
@automattic/full-site-editing: npm ERR! errno 1
@automattic/full-site-editing: npm ERR! @automattic/full-site-editing@0.18.2 test:e2e: `npx wp-scripts test-e2e --wordpress-base-url='http://localhost:9999'`
@automattic/full-site-editing: npm ERR! Exit status 1
@automattic/full-site-editing: npm ERR!
@automattic/full-site-editing: npm ERR! Failed at the @automattic/full-site-editing@0.18.2 test:e2e script.
```

I'll try to dig in more.

Great work getting us this far! 🎉

noahtallen commented:
Awesome, thanks @mattwiebe! 

> Navigation failed because browser has disconnected

I did also get this output early on, but I thought I had fixed it by rewriting the test a bit.... hm

> add npm ci step to the init script

awesome idea, we should definitely do that

noahtallen commented:
Hopefully this rebase clears up the tests? Really confused about why they're all messed up

Addison-Stavlo commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/tests/e2e/fse-back-button.spec.js:
```diff
@@ -0,0 +1,13 @@
+import { createNewPost } from '@wordpress/e2e-test-utils';
+
+describe( 'Full Site Editing Back Button', () => {
+	it( 'Should navigate to a new post.', async () => {
```
Im a little confused on what we are testing here.  Why should the FSE back button navigate to a new post?  

I am guessing `createNewPost()` leads us to that page, but we are just checking the title and not anything with the Back Button itself?  Is this test currently a placeholder to ensure the suite is running properly or something?

noahtallen replied:
Yes, it is just a placeholder :) Originally I was going to go into more depth, but I felt we could wait to go into depth after we merge this. So I should probably rename that

Addison-Stavlo reviewed (commented):
This is fantastic to see!

Everything seems to work well for me except I cant verify the e2e part currently.  
When I try to run that I get:
`Error: Chromium revision is not downloaded. Run "npm install"`, even after running "npm install"

Addison-Stavlo commented:
Hrm.. it seems after running this, every time I try to go back to my normal local WP environment on `localhost:8889`, it redirects to `localhost:9999`.  Even after running `npm run env start` back on my base gutenberg directory, it just automatically redirects to this new `localhost:9999` instead (which is no longer running at that point).  🤔   I wonder if there is some issue here?  Or maybe Im forgetting to do something...

noahtallen commented:
> Hrm.. it seems after running this, every time I try to go back to my normal local WP environment on localhost:8889

hm. That's very strange. I wonder if the docker setup uses the same name or ID or something, and then redirects the port. Honestly I'm not sure; I'll have to do some debugging myself.

> Error: Chromium revision is not downloaded. Run "npm install", even after running "npm install"

Ahhhhh that must be from https://github.com/Automattic/wp-calypso/pull/38900 which was merged after I started this PR. I'll fix that.

noahtallen commented:
I think to fix the chromium error, you need to run `npm ci` instead of `npm install`.

Addison-Stavlo commented:
Ok cool, I can get the e2e running with that.  It runs the suite, fails the test due to timeout, sums up the test results, and then exits with error code 1:
```
@automattic/full-site-editing:   Full Site Editing Back Button
@automattic/full-site-editing:     ✕ Should navigate to a new post. (5003ms)
@automattic/full-site-editing:   ● Full Site Editing Back Button › Should navigate to a new post.
@automattic/full-site-editing:     : Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Error:
@automattic/full-site-editing:        6 |
@automattic/full-site-editing:        7 | describe( 'Full Site Editing Back Button', () => {
@automattic/full-site-editing:     >  8 | 	it( 'Should navigate to a new post.', async () => {
@automattic/full-site-editing:          | 	^
@automattic/full-site-editing:        9 | 		await createNewPost();
@automattic/full-site-editing:       10 | 		const title = await page.title();
@automattic/full-site-editing:       11 | 		expect( title ).toBe( 'Add New Post ‹ WordPress Develop — WordPress' );
@automattic/full-site-editing:       at new Spec (../../node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
@automattic/full-site-editing:       at Suite.<anonymous> (full-site-editing-plugin/full-site-editing/tests/e2e/fse-back-button.spec.js:8:2)
@automattic/full-site-editing: Test Suites: 1 failed, 1 total
@automattic/full-site-editing: Tests:       1 failed, 1 total
@automattic/full-site-editing: Snapshots:   0 total
@automattic/full-site-editing: Time:        5.648s
@automattic/full-site-editing: Ran all test suites.
@automattic/full-site-editing: npm ERR! code ELIFECYCLE
@automattic/full-site-editing: npm ERR! errno 1
@automattic/full-site-editing: npm ERR! @automattic/full-site-editing@0.18.2 test:e2e: `npx wp-scripts test-e2e --wordpress-base-url='http://localhost:9999'`
@automattic/full-site-editing: npm ERR! Exit status 1
@automattic/full-site-editing: npm ERR!
@automattic/full-site-editing: npm ERR! Failed at the @automattic/full-site-editing@0.18.2 test:e2e script.
@automattic/full-site-editing: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
@automattic/full-site-editing: npm ERR! A complete log of this run can be found in:
@automattic/full-site-editing: npm ERR!     /Users/addisonstavlo/.npm/_logs/2020-01-28T23_42_42_199Z-debug.log
lerna ERR! npm run test:e2e exited 1 in '@automattic/full-site-editing'
```

But it is running the suite! 🎉

noahtallen commented:
see p4TIVU-9lk-p2 for info on the failing npm stuff. I'm trying to run npm several times in a row to generate a correct installation so that I can test using `npm ci`.

noahtallen commented:
> Ok cool, I can get the e2e running with that

great! looks like the setup is working ish but the tests need work, which makes sense :)

noahtallen commented:
Yeah, we really need to fix the local port redirect thing. I really don't understand that, but it's definitely a blocker. It completely messed up my local docker config

vindl commented:
Initially I was seeing 403 Forbidden error coming from Nginx while trying to access the Docker WP instance, and I'm guessing as a consequence the tests were also failing. I tried to debug it further but now it seems to be gone, and I haven't done anything special except go through the commands in testing instructions again.

All of the test suites are running fine for me now (php, js, e2e) and I'm noticing the same failing e2e test as @Addison-Stavlo and @mattwiebe have pointed out above.

Nice work @noahtallen! ✨

Addison-Stavlo commented:
> Yeah, we really need to fix the local port redirect thing. I really don't understand that, but it's definitely a blocker. It completely messed up my local docker config

I was able to 'fix' the redirect and my local gutenberg setup by deleting the wordpress folder there and running `npm run env install` again.  Unfortunately that will also nuke you local database if you have things there that you care about keeping.  `npm run env reinstall` and other things did not seem to work.

If we are setting these tests up to run on some remote docker container, I don't think we have an issue.  But running them locally, the wordpress install definitely causes some conflict with what is already in place.

noahtallen commented:
> But running them locally, the wordpress install definitely causes some conflict with what is already in place.

I think an acceptable solution would be to link the FSE setup to your existing gutenberg or WordPress setup if it already exists. I'll try to look into doing that. I know there is a "link" command built in which is supposed to accomplish that.

noahtallen commented:
just coming back to this:

> I was able to 'fix' the redirect and my local gutenberg setup by deleting the wordpress folder there and running npm run env install again. 

It also took me a long time to fix the redirect. For some reason, I feel like my browser was caching it way after Docker stopped doing it!

Here's my plan now:
- [ ] Fix the Docker setup if you have an existing thing
- [ ] Remove the e2e test so we can work on improving this aspect of the PR later.

noahtallen committed:
Create init script to test wp-env setup

noahtallen committed:
Fix paths in wp-env config and add support for dev version of wp-env

noahtallen commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/test/__snapshots__/template-selector-control.test.js.snap:
```diff
@@ -35,12 +35,6 @@ exports[`TemplateSelectorControl Basic rendering renders with required props 1`]
                 class="template-selector-item__media"
               />
             </div>
-            <span
```
This span was removed here: https://github.com/Automattic/wp-calypso/pull/39120/files#

noahtallen commented:
Current status:
- I have a local docker environment working with @wordpress/env, which you can test using the instructions above.
- Fixed js unit tests which started failing on this branch once the config got fixed (I think)
- Working on integrate e2e, phpunit, and phpcs to @wordpress/env on the core side of things so that we can continue using them here.

noahtallen committed:
Remove init script from package.json

noahtallen commented:
Update: see the new PR description. I've removed all the testing related code from this PR and narrowed it down to _just_ adding @wordpress/env. This should make it easier for me to iterate on the individual pieces of the puzzle :)

In particular, https://github.com/Automattic/wp-calypso/pull/39326 should be completely ready to go.

Still doing some work on this PR and the e2e one, but the phpunit one is completely blocked at the moment.

noahtallen committed:
Remove name from .wp-env.json file

Because it is not used for anything :)

noahtallen committed:
Add custom port numbers to .wp-env.json

@see https://github.com/WordPress/gutenberg/pull/20158

noahtallen committed:
Fix FSE build source

vindl commented on apps/full-site-editing/.wp-env.json:
```diff
@@ -0,0 +1,6 @@
+{
+	"plugins": [ "./full-site-editing-plugin", "../../../gutenberg" ],
```
What do you think about including Jetpack and wpcomsh here so it's closer to Atomic env?

noahtallen replied:
We could definitely do that. Do you think it could be a follow up?

One of the reasons that it might not be nice to do this is that every plugin and theme here is the development version. So I'd have to do `npm i` and `npm run build` for every plugin so that it loads correctly. And then I would have to make sure that I do `git pull` regularly. I could see that being a pain with a lot of plugins.

vindl replied:
> Do you think it could be a follow up?

Sure, that's fine with me.

vindl commented:
The current testing steps work fine for me up to a certain point. 🚀The site's front end and wp-admin load fine. When I try to open the editor though I'm running into this fatal, not sure how it's related given that this is a core WP function.

```
Fatal error: Uncaught Error: Call to undefined function wp_get_registered_image_subsizes() in /var/www/html/wp-content/plugins/gutenberg/lib/client-assets.php:690 Stack trace: #0 /var/www/html/wp-includes/class-wp-hook.php(288): gutenberg_extend_settings_image_dimensions(Array) #1 /var/www/html/wp-includes/plugin.php(208): WP_Hook->apply_filters(Array, Array) #2 /var/www/html/wp-admin/edit-form-blocks.php(374): apply_filters('block_editor_se...', Array, Object(WP_Post)) #3 /var/www/html/wp-admin/post.php(178): include('/var/www/html/w...') #4 {main} thrown in /var/www/html/wp-content/plugins/gutenberg/lib/client-assets.php on line 690
```

I'm guessing this is also the culprit behind my errors in e2e testing PR.

Also, what do you think about adding `./apps/full-site-editing/bin/setup-env.sh` as a script (e.g. `start-fse-env`)?

noahtallen commented:
> what do you think about adding ./apps/full-site-editing/bin/setup-env.sh as a script

Honestly, I wanted to do that, but I was thinking that it'd be really verbose to run it with npm:

`npx lerna run start-fse-env --scope='@automattic/full-site-editing' --stream --no-prefix"`

at which point referencing the script directly seems to make more sense. Plus, the script is written such that expects to start in the wp-calypso root, and not in the FSE app folder. (Though I could improve that :))

Plus, this is just for setting up the environment, so it should be a one-time thing. Once everything is working, you should be able to run `wp-env start` from the FSE root and it will work automatically.

> When I try to open the editor though I'm running into this fatal, not sure how it's related given that this is a core WP function.

My only guess is that it's a bug in Gutenberg?? The WordPress version is the latest production version (I think 5.3), not the dev version. Could try re-building the Gutenberg plugin to see if that works since the error is thrown in `plugins/gutenberg`.

vindl commented:
> Honestly, I wanted to do that, but I was thinking that it'd be really verbose to run it with npm:

Oh true, for a moment I forgot how much incantations it takes now. 😄

noahtallen committed:
Use new port syntax

vindl commented:
I tested this and everything works fine except loading the Gutenberg editor which results in the following fatal:

```
Fatal error: Uncaught Error: Call to undefined function wp_get_registered_image_subsizes() in /var/www/html/wp-content/plugins/gutenberg/lib/client-assets.php:690 Stack trace: #0 /var/www/html/wp-includes/class-wp-hook.php(288): gutenberg_extend_settings_image_dimensions(Array) #1 /var/www/html/wp-includes/plugin.php(208): WP_Hook->apply_filters(Array, Array) #2 /var/www/html/wp-admin/edit-form-blocks.php(374): apply_filters('block_editor_se...', Array, Object(WP_Post)) #3 /var/www/html/wp-admin/post-new.php(72): include('/var/www/html/w...') #4 {main} thrown in /var/www/html/wp-content/plugins/gutenberg/lib/client-assets.php on line 690
```

I tracked it down, and the problem seems to be stemming from the fact that my env is running WP 5.2.2, and required `wp_get_registered_image_subsizes` function that latest `master` of Gutenberg depends on was [introduced in WP 5.3](https://developer.wordpress.org/reference/functions/wp_get_registered_image_subsizes/#changelog).

I suspected that docker might be using some of my old WP images, but I purged them all and made sure that the script downloads a new one, and it still ended up being 5.2.2.

Having said that, this problem is not caused by this PR and it's something that needs to be resolved in the WP docker image or `wp-env` script.

vindl reviewed (approved):
LGTM! :shipit:
